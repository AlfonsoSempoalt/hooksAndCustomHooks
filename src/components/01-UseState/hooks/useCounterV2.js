import { useState } from 'react';

export const useCounter = (defaultValue = 0) => {
  const [counter, setCounter] = useState({defaultValue});

  const increase = () => {
    setCounter(setCounter(counter+1));
  };

  const decrease = () => {
    setCounter(setCounter(counter-1));
  };

  const restart = () => {
    setCounter(setCounter(defaultValue));
  };
 
  return {
    increase: increase,
    decrease: decrease,
    restart: restart,
    counter,
  };
};
