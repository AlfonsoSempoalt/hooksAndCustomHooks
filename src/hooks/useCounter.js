import { useState } from 'react';

export const useCounter = (defaultValue = 0) => {
  const [counter, setCounter] = useState({
    counter1: defaultValue,
    counter2: defaultValue,
  });

  const increaseCounter1 = (factor=1) => {
    setCounter({ ...counter, counter1: counter.counter1 + factor });
  };

  const decreaseCounter1 = (factor=1) => {
    setCounter({ ...counter, counter1: counter.counter1 - factor });
  };

  const restartCounter1 = () => {
    setCounter({ ...counter, counter1: defaultValue });
  };

  const increaseCounter2 = (factor=1) => {
    setCounter({ ...counter, counter2: counter.counter2 + factor });
  };

  const decreaseCounter2 = (factor=1) => {
    setCounter({ ...counter, counter2: counter.counter2 - factor });
  };

  const restartCounter2 = () => {
    setCounter({ ...counter, counter2: defaultValue });
  };

  return {
    increaseCounter1: increaseCounter1,
    decreaseCounter1: decreaseCounter1,
    restartCounter1: restartCounter1,
    increaseCounter2: increaseCounter2,
    decreaseCounter2: decreaseCounter2,
    restartCounter2: restartCounter2,
    counter,
  };
};
