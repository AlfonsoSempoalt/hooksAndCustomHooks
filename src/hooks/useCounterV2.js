import { useState } from 'react';

export const useCounter = (defaultValue = 1) => {
  const [counter, setCounter] = useState(defaultValue);

  const increase = () => {
    setCounter(counter+1);
  };

  const decrease = () => {
    setCounter(counter-1);
  };

  const restart = () => {
    setCounter(defaultValue);
  };
 
  return {
    increase: increase,
    decrease: decrease,
    restart: restart,
    counter: counter,
  };
};
