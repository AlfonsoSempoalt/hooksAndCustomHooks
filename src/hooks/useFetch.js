import { useEffect, useRef, useState } from 'react';

export const useFetch = (url) => {
    const isMounted = useRef(true);

    const [state, setState] = useState({
        data: null,
        loading: true,
        error: null,
    });

    useEffect(() => {
        return () => {
            isMounted.current = false;
        };
    }, []);

    useEffect(() => {
        setState({ data: null, loading: true, error: null });
        fetch(url)
            .then((response) => response.json())
            .then((data) => {
                if (isMounted.current) {
                    setState({ loading: false, data: data, error: null });
                }
            })
            .catch((error) => {
                setState({
                    data: null,
                    loading: false,
                    error: 'problems with the url petition',
                });
            });
    }, [url]);

    return state;
};
