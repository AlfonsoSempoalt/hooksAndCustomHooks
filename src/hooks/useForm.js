import { useState } from 'react';

export const useForm = (initialState = {}) => {
    const [formValues, setValues] = useState(initialState);

    const handleInputChange = ({ target }) => {
        setValues({ ...formValues, [target.name]: target.value });
    };

    const resetForm = () => {
        setValues(initialState);
    };

    const handleOnSubmit = (e) => {
        e.preventDefault();
    };

    return [formValues, 
            handleInputChange,
            resetForm,
            handleOnSubmit];
};
