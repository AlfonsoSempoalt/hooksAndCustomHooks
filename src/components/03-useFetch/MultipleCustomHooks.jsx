import React from 'react';
import { useCounter } from '../../hooks/useCounterV2';
import { useFetch } from '../../hooks/useFetch';
import '../02-UseEffect/effects.css';
const MultipleCustomHooks = () => {
  
  const { increase, counter } = useCounter(1);
  
  const { data, loading } = useFetch(
    `https://www.breakingbadapi.com/api/quotes/${counter}`
  );

  const { author, quote } = !!data && data[0];

  return (
    <>
      <h1>Custom hooks</h1>
      {loading ? (
        <section className='alert alert-info text-center'>loading...</section>
      ) : (
        <blockquote className='blockquote text-right'>
          <p className='mb-2'>{quote}</p>
          <footer className='blockquote-footer'>{author}</footer>
        </blockquote>
      )}

      <button className='btn btn-primary' onClick={increase}>
        Next
      </button>
    </>
  );
};

export default MultipleCustomHooks;
