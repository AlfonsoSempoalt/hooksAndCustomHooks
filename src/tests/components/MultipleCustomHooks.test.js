import React from 'react';
import '@testing-library/jest-dom';
import { expect, describe } from '@jest/globals';
import { shallow } from 'enzyme';
import MultipleCustomHooks from '../../components/03-useFetch/MultipleCustomHooks';
import { useFetch } from '../../hooks/useFetch';
jest.mock('../../hooks/useFetch');

describe('tests to MultipleCustomHooks.jsx', () => {
    test('should match with the snapshot', () => {
        useFetch.mockReturnValue({
            data: null,
            loading: true,
            error: null,
        });

        const wrapper = shallow(<MultipleCustomHooks />);
        expect(wrapper).toMatchSnapshot();
    });

    test('should charge the information', () => {
        useFetch.mockReturnValue({
            data: [
                {
                    author: 'Alfonso',
                    quote: 'Hello world',
                },
            ],
            loading: false,
            error: null,
        });

        const wrapper = shallow(<MultipleCustomHooks />);
        expect(wrapper.find('.alert').exists()).toBe(false);
        expect(wrapper.find('.mb-2').text().trim()).toBe('Hello world');
        expect(wrapper.find('.blockquote-footer').text().trim()).toBe(
            'Alfonso'
        );
    });
});
