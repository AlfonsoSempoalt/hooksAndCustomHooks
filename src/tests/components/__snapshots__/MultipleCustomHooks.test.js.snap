// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests to MultipleCustomHooks.jsx should match with the snapshot 1`] = `
<Fragment>
  <h1>
    Custom hooks
  </h1>
  <section
    className="alert alert-info text-center"
  >
    loading...
  </section>
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Next
  </button>
</Fragment>
`;
